### PR TITLE
feat(Computability): Add Chomsky Normal Form Grammar and translation

### PR DESCRIPTION
**Pull request by AlexLoitzl**
- Define Chomsky normal form grammars
- Add language-preserving translation between context-free grammars and Chomsky normal form grammars

Co-authored-by: Martin Dvorak martin.dvorak@matfyz.cz 

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Computability/ContextFreeGrammar.lean [hunk 1]
```diff
@@ -194,6 +207,13 @@ lemma Derives.append_right {v w : List (Symbol T g.NT)}
   | refl => rfl
   | tail _ last ih => exact ih.trans_produces <| last.append_right p
 
+theorem Derives.head_induction_on {v : List (Symbol T g.NT)} {P : ∀ u, g.Derives u v → Prop}
+  {u : List (Symbol T g.NT)} (huv : g.Derives u v)
```
Comment (madvorak) [inline Mathlib/Computability/ContextFreeGrammar.lean:211]: indent four spaces please

File: Mathlib/Computability/ChomskyNormalForm/Basic.lean [hunk 2]
```diff
@@ -0,0 +1,494 @@
...
+/-- Rule `r` rewrites string `u` to string `v` iff they share both a prefix `p` and postfix `q`
+such that the remaining middle part of `u` is the input of `r` and the remaining middle part
+of `v` is the output of `r`. -/
+theorem rewrites_iff {r : ChomskyNormalFormRule T N} (u v : List (Symbol T N)) :
+    r.Rewrites u v ↔ ∃ p q : List (Symbol T N),
+      u = p ++ [Symbol.nonterminal r.input] ++ q ∧ v = p ++ r.output ++ q := by
+  constructor
+  · apply Rewrites.exists_parts
+  · rintro ⟨p, q, rfl, rfl⟩
+    apply rewrites_of_exists_parts
```
Comment (YaelDillies) [inline Mathlib/Computability/ChomskyNormalForm/Basic.lean:108]: Since `Iff` is a structure
```suggestion
      u = p ++ [Symbol.nonterminal r.input] ++ q ∧ v = p ++ r.output ++ q where
  mp := Rewrites.exists_parts
  mpr := by rintro ⟨p, q, rfl, rfl⟩; apply rewrites_of_exists_parts
```

File: Mathlib/Computability/ChomskyNormalForm/Basic.lean [hunk 3]
```diff
@@ -0,0 +1,494 @@
...
+lemma Rewrites.input_output : r.Rewrites [.nonterminal r.input] r.output := by
+  cases r
+  · simpa using head_leaf _ _ []
+  · simpa using head_node _ _ _ []
+
+lemma rewrites_of_exists_parts (r : ChomskyNormalFormRule T N) (p q : List (Symbol T N)) :
+    r.Rewrites (p ++ [Symbol.nonterminal r.input] ++ q) (p ++ r.output ++ q) := by
+    induction p with
+    | nil => cases r <;> constructor
+    | cons _ _ hr => exact Rewrites.cons r _ hr
```
Comment (YaelDillies) [inline Mathlib/Computability/ChomskyNormalForm/Basic.lean:97]:
```suggestion
  induction p with
  | nil => cases r <;> constructor
  | cons _ _ hr => exact .cons r _ hr
```

File: Mathlib/Computability/ChomskyNormalForm/Basic.lean [hunk 4]
```diff
@@ -0,0 +1,494 @@
...
+    rcases hrs with ⟨p, q, rfl, rfl⟩
+    use x :: p, q
+    simp
+
+lemma Rewrites.input_output : r.Rewrites [.nonterminal r.input] r.output := by
+  cases r
+  · simpa using head_leaf _ _ []
+  · simpa using head_node _ _ _ []
+
+lemma rewrites_of_exists_parts (r : ChomskyNormalFormRule T N) (p q : List (Symbol T N)) :
```
Comment (YaelDillies) [inline Mathlib/Computability/ChomskyNormalForm/Basic.lean:93]:
```suggestion
lemma Rewrites.of_exists_parts (r : ChomskyNormalFormRule T N) (p q : List (Symbol T N)) :
```
for anonymous dot notation

File: Mathlib/Computability/ChomskyNormalForm/Basic.lean [hunk 5]
```diff
@@ -0,0 +1,494 @@
...
+of `v` is the output of `r`. -/
+theorem rewrites_iff {r : ChomskyNormalFormRule T N} (u v : List (Symbol T N)) :
+    r.Rewrites u v ↔ ∃ p q : List (Symbol T N),
+      u = p ++ [Symbol.nonterminal r.input] ++ q ∧ v = p ++ r.output ++ q := by
+  constructor
+  · apply Rewrites.exists_parts
+  · rintro ⟨p, q, rfl, rfl⟩
+    apply rewrites_of_exists_parts
+
+/-- Add extra prefix to cnf rewriting. -/
```
Comment (YaelDillies) [inline Mathlib/Computability/ChomskyNormalForm/Basic.lean:110]:
```suggestion
/-- Add extra prefix to CNF rewriting. -/
```

File: Mathlib/Computability/ChomskyNormalForm/Basic.lean [hunk 6]
```diff
@@ -0,0 +1,494 @@
...
+      · exact hw
+      · constructor
+        · apply Produces.trans_derivesIn
+          · use r
+            constructor
+            · exact hrg
+            · rw [hp, ← List.singleton_append, ← List.append_assoc]
+              apply r.rewrites_of_exists_parts
+          · exact hd₁
+        · exact ⟨hd₂, by omega⟩
```
Comment (YaelDillies) [inline Mathlib/Computability/ChomskyNormalForm/Basic.lean:456]: It's good style to not spend too many lines moving data around when it is done in a straightforward way
```suggestion
      refine ⟨x, y, m₁ + 1, m₂, hw, Produces.trans_derivesIn ⟨r, hrg, ?_⟩ hd₁, hd₂, by omega⟩
      rw [hp, ← List.singleton_append, ← List.append_assoc]
      apply r.rewrites_of_exists_parts
```

**Review by YaelDillies — COMMENTED** [review]
Could you please act on the comments below everywhere they apply (ie not just where I've commented)?

File: Mathlib/Computability/ChomskyNormalForm/Basic.lean (hunk 2 above)
Comment (madvorak) [reply Mathlib/Computability/ChomskyNormalForm/Basic.lean:108]: I want a second opinion. I have never written `mp` and `mpr` explicitly in my life.

**Comment by kim-em** [comment]
Thanks, @YaelDillies, for the reviews.

At the same time, I'll note that internal reviewer discussion currently points to us deciding that this is out of scope for Mathlib, with insufficient maintainer/reviewer expertise. (Note it's not just reviewing the material initially, but being confident we have expertise if this need repairing or refactoring in a few years.) No definitive answer was reached, and Yael's review may push things in a different direction. But I thought it good to be up front about the discussion.

**Comment by YaelDillies** [comment]
My opinion is that the current PR is too big and has some style issues, but that if I review it could become mergeable and you might learn something 😄

**Comment by AlexLoitzl** [comment]
Thank you @YaelDillies!
It's my first time working with Lean and the PR is a building block (although much larger than the actual proof) for proving the Pumping Lemma for CFG, which i have already finished [here](https://github.com/AlexLoitzl/pumping_cfg).

Regarding the size, I could not think of a reasonable splitting of the PR, since I thought it was nonsensical to split it into multiple PRs which do some translation of CFG with no clear goal.

**Comment by YaelDillies** [comment]
Being merged or not being merged I think is not greatly relevant. We are at a pivotal period of mathlib where we're hitting the limits of the monorepo model. Instead of accepting everything blindly in the `mathlib4` repo, we are now considering the possibility of having several mathlib repos, a central one (this one) along with one for each "mathematical dead end", namely one for each subdomain whose results are unlikely to ever be used in another domain. Eg here we could have the repo `mathlib-formal-language`. See https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/Repository.20for.20combinatorial.20games for more information.

Long story short, I think it's worth reviewing anyway.

File: Mathlib/Computability/ChomskyNormalForm/Basic.lean (hunk 4 above)
Comment (AlexLoitzl) [reply Mathlib/Computability/ChomskyNormalForm/Basic.lean:93]: I either don't understand why this should change or how to use it properly. If I change it as suggested, I need to spell out the entire Lemma whenever I want to use it as there is no parameter of type `Rewrites`. Hence, instead of writing `r.rewrites_of_exists_parts`, I now find myself writing `ChomskyNormalFormRule.Rewrites.of_exists_parts`. I picked the name similar to the same lemma `rewrites_of_exists_parts` of ContextFreeGrammar.

Comment (YaelDillies) [reply Mathlib/Computability/ChomskyNormalForm/Basic.lean:93]: You mustn't have encountered anonymous dot notation yet. Wherever you have a goal of the form `⊢ Rewrites _ _ _`, you can now do `exact .of_exists_parts _ _ _ _`

Comment (AlexLoitzl) [reply Mathlib/Computability/ChomskyNormalForm/Basic.lean:93]: Nice, thank you!
I have seen it, but tried using it with `apply .of_exists_parts` and thought I am missing something

Comment (YaelDillies) [reply Mathlib/Computability/ChomskyNormalForm/Basic.lean:93]: Ah yes, `apply` doesn't elaborate using the expected type, so can't use anonymous dot notation

File: Mathlib/Computability/ChomskyNormalForm/Basic.lean (hunk 2 above)
Comment (AlexLoitzl) [reply Mathlib/Computability/ChomskyNormalForm/Basic.lean:108]: Is this true in general, or only if the terms can be stated concisely, in particular, should I structure a long proof as
````
lemma ... where
mp := by ...
mpr := by ...
````
vs.,
```
lemma ... := by
constructor
· ...
· ...

Comment (YaelDillies) [reply Mathlib/Computability/ChomskyNormalForm/Basic.lean]: I would say if the terms can be stated concisely, you should do it. Otherwise, it's up to you

File: Mathlib/Computability/ChomskyNormalForm/Basic.lean (hunk 6 above)
Comment (AlexLoitzl) [reply Mathlib/Computability/ChomskyNormalForm/Basic.lean]: I tried my best... I am mostly struggling if rewrites/intros are involved, where I don't know if/how I can avoid breaking things apart.

File: Mathlib/Computability/ChomskyNormalForm/LengthRestriction.lean [hunk 7]
```diff
@@ -83,11 +83,9 @@ lemma Wellformed.cases {r : ContextFreeRule T N} (hr : r.Wellformed) :
     | .terminal t :: _ :: _ => specialize hu (Symbol.terminal t); simp at hu
     | _ :: .terminal t :: _ => specialize hu (Symbol.terminal t); simp at hu
     | .nonterminal n₁ :: .nonterminal n₂ :: u =>
-      right
       simp only [List.mem_cons, forall_eq_or_imp, true_and] at hu
       obtain ⟨u', huu⟩ := only_nonterminals hu
-      use n₁, n₂, u'
-      simp [huu.symm]
+      exact Or.inr ⟨n₁, n₂, u', huu ▸ rfl⟩
```
Comment (YaelDillies) [inline Mathlib/Computability/ChomskyNormalForm/LengthRestriction.lean:88]:
```suggestion
      exact .inr ⟨n₁, n₂, u', huu ▸ rfl⟩
```

File: Mathlib/Computability/ChomskyNormalForm/LengthRestriction.lean [hunk 8]
```diff
@@ -368,8 +363,7 @@ lemma restrictLength_produces_derives_projectString {u v : List (Symbol T g.NT')
   simp only [restrictLength] at r hrg
   rw [hu, hv]
   repeat rw [projectString_append]
-  apply Derives.append_right
-  apply Derives.append_left
+  refine Derives.append_right (Derives.append_left ?_ _) _
```
Comment (YaelDillies) [inline Mathlib/Computability/ChomskyNormalForm/LengthRestriction.lean:366]:
```suggestion
  refine .append_right (.append_left ?_ _) _
```
If `append_right`/`append_left` had one less explicit argument, you could even write
```suggestion
  refine .append_right <| .append_left ?_
```

File: Mathlib/Computability/ChomskyNormalForm/TerminalRestriction.lean [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+  induction u with
+  | nil => rfl
+  | cons a =>
+    simp only [rightEmbedString, projectString, List.map_cons, List.map_map, List.cons.injEq]
+    constructor
+    · cases a <;> rfl
+    · rwa [← List.map_map]
+
+lemma projectString_terminals {u : List T} :
+    projectString (u.map (@Symbol.terminal T (N ⊕ T))) = u.map Symbol.terminal := by
```
Comment (YaelDillies) [inline Mathlib/Computability/ChomskyNormalForm/TerminalRestriction.lean:82]: There's nothing called `terminals` here. Did you mean
```suggestion
lemma projectString_terminal {u : List T} :
    projectString (u.map (@Symbol.terminal T (N ⊕ T))) = u.map Symbol.terminal := by
```

File: Mathlib/Computability/ChomskyNormalForm/TerminalRestriction.lean [hunk 10]
```diff
@@ -0,0 +1,320 @@
...
+/-! Definitions of embeding into and projecting to the type of symbols of the new grammar -/
+section EmbedProject
+
+variable {N : Type uN}
+
+/-- Intuitive embedding of symbols of the original grammar into symbols of the new grammar's type -/
+def embedSymbol (s : Symbol T N) : Symbol T (N ⊕ T) :=
+  match s with
+  | Symbol.terminal t => Symbol.terminal t
+  | Symbol.nonterminal n => Symbol.nonterminal (Sum.inl n)
```
Comment (YaelDillies) [inline Mathlib/Computability/ChomskyNormalForm/TerminalRestriction.lean:42]:
```suggestion
  | .terminal t => .terminal t
  | .nonterminal n => .nonterminal (.inl n)
```
Here and everywhere else

File: Mathlib/Computability/ChomskyNormalForm/TerminalRestriction.lean [hunk 11]
```diff
@@ -0,0 +1,320 @@
...
+variable {T : Type uT}
+
+/-! Definitions of embeding into and projecting to the type of symbols of the new grammar -/
+section EmbedProject
+
+variable {N : Type uN}
+
+/-- Intuitive embedding of symbols of the original grammar into symbols of the new grammar's type -/
+def embedSymbol (s : Symbol T N) : Symbol T (N ⊕ T) :=
+  match s with
```
Comment (YaelDillies) [inline Mathlib/Computability/ChomskyNormalForm/TerminalRestriction.lean:40]:
```suggestion
def embedSymbol : Symbol T N → Symbol T (N ⊕ T)
```
Here and everywhere else

File: Mathlib/Computability/ChomskyNormalForm/TerminalRestriction.lean [hunk 12]
```diff
@@ -0,0 +1,320 @@
...
+  | Symbol.terminal t => Symbol.terminal t
+  | Symbol.nonterminal (Sum.inl nt) => Symbol.nonterminal nt
+  | Symbol.nonterminal (Sum.inr t) => Symbol.terminal t
+
+/-- Projection from strings of the new grammars type into strings of the original grammar -/
+def projectString (u : List (Symbol T (N ⊕ T))) : List (Symbol T N) := u.map projectSymbol
+
+lemma embedSymbol_nonterminal_eq {nt : N} :
+    embedSymbol (Symbol.nonterminal nt) = (@Symbol.nonterminal T (N ⊕ T)) (Sum.inl nt) := by
+  rfl
```
Comment (YaelDillies) [inline Mathlib/Computability/ChomskyNormalForm/TerminalRestriction.lean:69]:
```suggestion
@[simp]
lemma embedSymbol_nonterminal (nt : N) :
    embedSymbol (Symbol.nonterminal nt) = (@Symbol.nonterminal T (N ⊕ T)) (Sum.inl nt) := by
  rfl
```

File: Mathlib/Computability/ChomskyNormalForm/TerminalRestriction.lean (hunk 9 above)
Comment (YaelDillies) [inline Mathlib/Computability/ChomskyNormalForm/TerminalRestriction.lean:82]: You can do
```suggestion
lemma projectString_terminals {u : List T} :
    projectString (u.map .terminal : List (Symbol T (N ⊕ T))) = u.map Symbol.terminal := by
```
which is arguably more readable

File: Mathlib/Computability/ChomskyNormalForm/Translation.lean [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+# Chomsky Normal Form Translation
+
+This file contains the algorithm to translate a `ContextFreeGrammar.` to a
+`ChomskyNormalFormGrammar` by using the algorithms
+`ContextFreeGrammar.eliminateEmpty` removing rules with an empty right-hand side,
+`ContextFreeGrammar.eliminateUnitRules` removing rules with a single nonterminal right-hand side,
+`ContextFreeGrammar.restrictTerminals` restricting all terminals to only occur as the single
+symbol of a rule's right-hand side, and
+`ContextFreeGrammar.restrictLength` translating `ContextFreeRule`s with multiple nonterminals to
+`ChomskyNormalFormRule`s with only two nonterminals on the right-hand side.
```
Comment (YaelDillies) [inline Mathlib/Computability/ChomskyNormalForm/Translation.lean:24]: Make it a list maybe?
```suggestion
* `ContextFreeGrammar.eliminateEmpty` removing rules with an empty right-hand side,
* `ContextFreeGrammar.eliminateUnitRules` removing rules with a single nonterminal right-hand side,
* `ContextFreeGrammar.restrictTerminals` restricting all terminals to only occur as the single
  symbol of a rule's right-hand side, and
* `ContextFreeGrammar.restrictLength` translating `ContextFreeRule`s with multiple nonterminals to
  `ChomskyNormalFormRule`s with only two nonterminals on the right-hand side.
```

File: Mathlib/Computability/ContextFreeGrammar.lean [hunk 14]
```diff
@@ -136,6 +137,17 @@ lemma mem_language_iff (g : ContextFreeGrammar.{uN} T) (w : List T) :
 
 variable {g : ContextFreeGrammar.{uN} T}
 
+lemma Produces.input_output {r : ContextFreeRule T g.NT} (hrg : r ∈ g.rules) :
+    g.Produces [.nonterminal r.input] r.output := ⟨r, hrg, .input_output⟩
+
+lemma Produces.rule {n : g.NT} {u : List (Symbol T g.NT)}
```
Comment (YaelDillies) [inline Mathlib/Computability/ContextFreeGrammar.lean:143]:
```suggestion
lemma Produces.mem_rules {n : g.NT} {u : List (Symbol T g.NT)}
```

File: Mathlib/Computability/ContextFreeGrammar.lean [hunk 15]
```diff
@@ -194,6 +206,13 @@ lemma Derives.append_right {v w : List (Symbol T g.NT)}
   | refl => rfl
   | tail _ last ih => exact ih.trans_produces <| last.append_right p
 
+theorem Derives.head_induction_on {v : List (Symbol T g.NT)} {P : ∀ u, g.Derives u v → Prop}
+    {u : List (Symbol T g.NT)} (huv : g.Derives u v)
+    (refl : P v (Derives.refl v))
+    (head : ∀ {u w} (huw : g.Produces u w) (hwv : g.Derives w v), P w hwv → P u (hwv.head huw)) :
+    P u huv :=
+  Relation.ReflTransGen.head_induction_on huv refl head
```
Comment (YaelDillies) [inline Mathlib/Computability/ContextFreeGrammar.lean:214]:
```suggestion
nonrec theorem Derives.head_induction_on {v : List (Symbol T g.NT)} {P : ∀ u, g.Derives u v → Prop}
    {u : List (Symbol T g.NT)} (huv : g.Derives u v)
    (refl : P v (Derives.refl v))
    (head : ∀ {u w} (huw : g.Produces u w) (hwv : g.Derives w v), P w hwv → P u (hwv.head huw)) :
    P u huv :=
  huv.head_induction_on refl head
```

File: Mathlib/Computability/ContextFreeGrammar.lean [hunk 16]
```diff
@@ -299,3 +318,169 @@ theorem Language.IsContextFree.reverse (L : Language T) :
     L.IsContextFree → L.reverse.IsContextFree := by rintro ⟨g, rfl⟩; exact ⟨g.reverse, by simp⟩
 
 end closure_reversal
+
+/-! Alternative definition of `ContextFreeGrammar.Derives` which allows to use well-founded
+induction on derivations, by explicitely counting the number of steps of the transformation -/
```
Comment (YaelDillies) [inline Mathlib/Computability/ContextFreeGrammar.lean:323]: Do you want to make these headers?
```suggestion
/-! ### Alternative definition of `ContextFreeGrammar.Derives` which allows to use well-founded
induction on derivations, by explicitely counting the number of steps of the transformation -/
```

File: Mathlib/Computability/ChomskyNormalForm/LengthRestriction.lean (hunk 8 above)
Comment (AlexLoitzl) [reply Mathlib/Computability/ChomskyNormalForm/LengthRestriction.lean:366]: Regarding the second comment, it's probably best to keep it consistent with the corresponding definition for context-free grammars?

Comment (YaelDillies) [reply Mathlib/Computability/ChomskyNormalForm/LengthRestriction.lean:366]: > it's probably best to keep it consistent with the corresponding definition for context-free grammars?

Given that the context-free grammar API hasn't been battle-tested, I would say consistency is fine to break here. The better option would be to fix the corresponding context-free grammar definition

Comment (madvorak) [reply Mathlib/Computability/ChomskyNormalForm/LengthRestriction.lean:366]: You want to change a lemma, not a definition, right?

Comment (YaelDillies) [reply Mathlib/Computability/ChomskyNormalForm/LengthRestriction.lean:366]: Yes, it's a lemma

Comment (madvorak) [reply Mathlib/Computability/ChomskyNormalForm/LengthRestriction.lean:366]: I am not a fan of arguments being inferred from the output, but I will accept this change for `ContextFreeGrammar.{Produces|Derives}.append_{left|right}` itself.

File: Mathlib/Computability/ChomskyNormalForm/TerminalRestriction.lean (hunk 10 above)
Comment (AlexLoitzl) [reply Mathlib/Computability/ChomskyNormalForm/TerminalRestriction.lean:42]: Regarding the multiple comments about anonymous dot notation. Is there any reason not to use it, e.g. for clarity?
Otw I'll go though everything changing it whenever possible.

Comment (YaelDillies) [reply Mathlib/Computability/ChomskyNormalForm/TerminalRestriction.lean:42]: You can always hover over an anonymous dot notation use to see what it really refers to, and my personal philosophy is that clarity is *gained* when you use anonymous dot notation, because it signals "This is just the obvious constructor"

File: Mathlib/Computability/ChomskyNormalForm/Basic.lean [hunk 17]
```diff
@@ -197,35 +194,31 @@ lemma Derives.eq_or_head {u w : List (Symbol T g.NT)} (huw : g.Derives u w) :
...
 
 /-- Add extra prefix to CNF deriving. -/
-lemma Derives.append_left {u v : List (Symbol T g.NT)}
-    (huv : g.Derives u v) (p : List (Symbol T g.NT)) :
+lemma Derives.append_left {u v p : List (Symbol T g.NT)} (huv : g.Derives u v) :
     g.Derives (p ++ u) (p ++ v) := by
   induction huv with
   | refl => rfl
-  | tail _ last ih => exact ih.trans_produces <| last.append_left p
+  | tail _ last ih => exact ih.trans_produces <| last.append_left
```
Comment (YaelDillies) [inline Mathlib/Computability/ChomskyNormalForm/Basic.lean:214]:
```suggestion
  | tail _ last ih => exact ih.trans_produces last.append_left
```
and below

**Comment by YaelDillies** [comment]
Can you resolve the conversations which you've dealt with, and leave unresolved the ones to which you think I might want to add something?

File: Mathlib/Computability/ContextFreeGrammar.lean (hunk 15 above)
Comment (AlexLoitzl) [reply Mathlib/Computability/ContextFreeGrammar.lean]: I did this only here as I could not find a good explanation of the exact meaning  - or a definite answer on when to use it.

Comment (YaelDillies) [reply Mathlib/Computability/ContextFreeGrammar.lean]: `nonrec theorem foo`:= ...` means that `foo` is nonrecursive, so if you use something called `foo` inside the `...` Lean will know that it must be referring to something else than the thing you're proving. Here, `nonrec` lets Lean know that `.head_induction_on` doesn't refer to `Derives.head_induction_on`, so it must refer to `Relation.ReflTransGen.head_induction_on`

File: Mathlib/Computability/ChomskyNormalForm/Basic.lean (hunk 6 above)
Comment (YaelDillies) [reply Mathlib/Computability/ChomskyNormalForm/Basic.lean]: As general rule of thumb,:
* `use` and `constructor` can always be combined into a `refine`
* `exact` and `constructor` can always be combined into a `refine`
* `apply` leads to weak unification and I encourage you to use `refine` instead

File: Mathlib/Computability/ContextFreeGrammar.lean [hunk 18]
```diff
@@ -136,6 +137,17 @@ lemma mem_language_iff (g : ContextFreeGrammar.{uN} T) (w : List T) :
 
 variable {g : ContextFreeGrammar.{uN} T}
 
+lemma Produces.input_output {r : ContextFreeRule T g.NT} (hrg : r ∈ g.rules) :
+    g.Produces [.nonterminal r.input] r.output := ⟨r, hrg, .input_output⟩
+
+lemma Produces.mem_rule {n : g.NT} {u : List (Symbol T g.NT)}
```
Comment (YaelDillies) [inline Mathlib/Computability/ContextFreeGrammar.lean:143]:
```suggestion
lemma Produces.mem_rules {n : g.NT} {u : List (Symbol T g.NT)}
```

File: Mathlib/Computability/ContextFreeGrammar.lean [hunk 19]
```diff
@@ -136,6 +137,17 @@ lemma mem_language_iff (g : ContextFreeGrammar.{uN} T) (w : List T) :
...
+    (hnu : g.Produces [Symbol.nonterminal n] u) :
```
Comment (YaelDillies) [inline Mathlib/Computability/ContextFreeGrammar.lean:144]:
```suggestion
    (hnu : g.Produces [.nonterminal n] u) :
```

File: Mathlib/Computability/ContextFreeGrammar.lean [hunk 20]
```diff
@@ -299,3 +318,169 @@ theorem Language.IsContextFree.reverse (L : Language T) :
...
+`g.DerivesIn u v n` means that `g` can transform `u` to `v` in `n` rewriting steps. -/
+inductive DerivesIn (g : ContextFreeGrammar.{uN} T) :
+    List (Symbol T g.NT) → List (Symbol T g.NT) → ℕ → Prop
+  /-- 0 steps entail no transformation -/
+  | refl (w : List (Symbol T g.NT)) : g.DerivesIn w w 0
+  /-- n + 1 steps, if transforms `u` to `v` in n steps, and `v` to `w` in 1 step  -/
+  | tail (u v w : List (Symbol T g.NT)) (n : ℕ) :
+    g.DerivesIn u v n → g.Produces v w → g.DerivesIn u w n.succ
+
+lemma derives_iff_derivesIn (g : ContextFreeGrammar.{uN} T) (v w : List (Symbol T g.NT)) :
```
Comment (YaelDillies) [inline Mathlib/Computability/ContextFreeGrammar.lean:338]:
```suggestion
lemma derives_iff_exists_derivesIn {g : ContextFreeGrammar.{uN} T}
    {v w : List (Symbol T g.NT)} :
```

File: Mathlib/Computability/ContextFreeGrammar.lean [hunk 21]
```diff
@@ -299,3 +318,169 @@ theorem Language.IsContextFree.reverse (L : Language T) :
...
+      · exact ihn
+      · exact last
+  · intro ⟨n, hgvwn⟩
+    induction hgvwn with
+    | refl => rfl
+    | tail _ _ _ _ last ih => exact ih.trans_produces last
+
+lemma mem_language_iff_derivesIn (g : ContextFreeGrammar.{uN} T) (w : List T) :
+    w ∈ g.language ↔ ∃ n, g.DerivesIn [Symbol.nonterminal g.initial] (w.map Symbol.terminal) n := by
+  rw [mem_language_iff, derives_iff_derivesIn]
```
Comment (YaelDillies) [inline Mathlib/Computability/ContextFreeGrammar.lean:359]: No point duplicating the `Derives` API
```suggestion
```

File: Mathlib/Computability/ContextFreeGrammar.lean [hunk 22]
```diff
@@ -299,3 +318,169 @@ theorem Language.IsContextFree.reverse (L : Language T) :
...
+
+/-- Given a context-free grammar `g`, strings `u` and `v`, and number `n`
+`g.DerivesIn u v n` means that `g` can transform `u` to `v` in `n` rewriting steps. -/
+inductive DerivesIn (g : ContextFreeGrammar.{uN} T) :
+    List (Symbol T g.NT) → List (Symbol T g.NT) → ℕ → Prop
+  /-- 0 steps entail no transformation -/
+  | refl (w : List (Symbol T g.NT)) : g.DerivesIn w w 0
+  /-- n + 1 steps, if transforms `u` to `v` in n steps, and `v` to `w` in 1 step  -/
+  | tail (u v w : List (Symbol T g.NT)) (n : ℕ) :
+    g.DerivesIn u v n → g.Produces v w → g.DerivesIn u w n.succ
```
Comment (YaelDillies) [inline Mathlib/Computability/ContextFreeGrammar.lean:336]: I'd suggest
```suggestion
  | zero {w : List (Symbol T g.NT)} : g.DerivesIn w w 0
  /-- n + 1 steps, if transforms `u` to `v` in n steps, and `v` to `w` in 1 step  -/
  | succ {u v w : List (Symbol T g.NT)} {n : ℕ} :
    g.DerivesIn u v n → g.Produces v w → g.DerivesIn u w n.succ
```

File: Mathlib/Computability/ContextFreeGrammar.lean [hunk 23]
```diff
@@ -299,3 +318,169 @@ theorem Language.IsContextFree.reverse (L : Language T) :
...
+
+lemma derives_iff_derivesIn (g : ContextFreeGrammar.{uN} T) (v w : List (Symbol T g.NT)) :
+    g.Derives v w ↔ ∃ n : ℕ, g.DerivesIn v w n := by
+  constructor
+  · intro hgvw
```
Comment (YaelDillies) [inline Mathlib/Computability/ContextFreeGrammar.lean:341]:
```suggestion
    g.Derives v w ↔ ∃ n : ℕ, g.DerivesIn v w n where
  mp hgvw := by
```

File: Mathlib/Computability/ContextFreeGrammar.lean [hunk 24]
```diff
@@ -299,3 +318,169 @@ theorem Language.IsContextFree.reverse (L : Language T) :
...
+    | refl => rfl
+    | tail _ _ _ _ last ih => exact ih.trans_produces last
+
+lemma mem_language_iff_derivesIn (g : ContextFreeGrammar.{uN} T) (w : List T) :
+    w ∈ g.language ↔ ∃ n, g.DerivesIn [Symbol.nonterminal g.initial] (w.map Symbol.terminal) n := by
+  rw [mem_language_iff, derives_iff_derivesIn]
+
+variable {g : ContextFreeGrammar.{uN} T}
+
+lemma DerivesIn.zero_steps (w : List (Symbol T g.NT)) : g.DerivesIn w w 0 := refl w
```
Comment (YaelDillies) [inline Mathlib/Computability/ContextFreeGrammar.lean:363]: ... then
```suggestion
```

File: Mathlib/Computability/ContextFreeGrammar.lean [hunk 25]
```diff
@@ -299,3 +318,169 @@ theorem Language.IsContextFree.reverse (L : Language T) :
...
+
+lemma Produces.single_step {v w : List (Symbol T g.NT)} (hvw : g.Produces v w) :
```
Comment (YaelDillies) [inline Mathlib/Computability/ContextFreeGrammar.lean:365]:
```suggestion
lemma Produces.derivesIn_one {v w : List (Symbol T g.NT)} (hvw : g.Produces v w) :
```

File: Mathlib/Computability/ContextFreeGrammar.lean [hunk 26]
```diff
@@ -299,3 +318,169 @@ theorem Language.IsContextFree.reverse (L : Language T) :
...
+    g.DerivesIn v w 1 := DerivesIn.tail _ _ _ _ (DerivesIn.zero_steps v) hvw
```
Comment (YaelDillies) [inline Mathlib/Computability/ContextFreeGrammar.lean:366]:
```suggestion
    g.DerivesIn v w 1 := .succ .zero hvw
```

File: Mathlib/Computability/ContextFreeGrammar.lean
Comment (YaelDillies) [inline Mathlib/Computability/ContextFreeGrammar.lean:1]: Can you make this a separate PR? I think it's close to completion.

**Comment by joneugster** [comment]
Mathlib has moved to PRs from forks. This PR is still from a branch of the main repository and will soon be closed! Please migrate the content to a fork and reopen a PR from there if you wish to do so. See [Zulip topic](https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/all.20future.20PRs.20should.20be.20made.20from.20forks/with/624081853) for more instructions. Thank you for contributing to mathlib!
